### ISSUE TITLE
Create JwtAuthenticationToken with custom claims

### ISSUE THREAD

**Issue by wimdeblauwe** [ch4mpy/spring-addons#27]
I am currently using a custom annotation in my tests that builds op `@WithMockOidcId()` like this:

```
import com.c4_soft.springaddons.security.oauth2.test.annotations.ClaimSet;
import com.c4_soft.springaddons.security.oauth2.test.annotations.StringClaim;
import com.c4_soft.springaddons.security.oauth2.test.annotations.WithMockAuthentication;
import com.c4_soft.springaddons.security.oauth2.test.annotations.WithMockOidcId;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

@Retention(RetentionPolicy.RUNTIME)
@WithMockOidcId(authorities = "ROLE_ADMIN",
        privateClaims = @ClaimSet(stringClaims = @StringClaim(name = "myPrivateClaim", value = "xxxxxx xxxxxx")))
public @interface WithMockCasefileAdmin {
}
```

However, I would need that a `JwtAuthenticationToken` is created (I am using Keycloak, but I don't want to depend on Keycloak specific classes in my Spring Boot application).

I know that I can create such a token via:

```
@WithMockAuthentication(value = JwtAuthenticationToken.class, authorities = "ROLE_ADMIN")
```

But I can't configure claims in the annotation. Would it be possible to add this option?

**Comment by ch4mpy** [comment]
Hello @wimdeblauwe ,

The `@WithMockOidcId` builds and `OidcIdToken` which is not tied to Keycloak. I already used it to test apps with Auth0 or Microsoft IdentityServer as authorisation server.

`JwtAuthenticationToken` interface is very poor: all you can get is a `Map<String, Object>` through `getTokenAttributes()`. In my opinion, it makes it hardly usable.

This `OidcIdToken` I mentioned is exposing all standard claims from OpenID and is vendor neutral (no auto mapping for vendors private claims). All you have to do is providing a converter responsible of figuring out the authorities based on the info in the token (and whatever else is relevant).

You can find in my [starter](https://github.com/ch4mpy/starter) project, sample configuration for both [servlet](https://github.com/ch4mpy/starter/tree/master/api/webmvc/common-security-webmvc/src/main/java/com/c4_soft/commons/security) and [reactive](https://github.com/ch4mpy/starter/tree/master/api/webflux/common-security-webflux/src/main/java/com/c4_soft/commons/security) apps.

To answer your question, I see two ways to have custom `JwtAuthenticationToken` in test security context, but neither is exiting:
- copy from `@WithMockOidcId` and build your own annotation
- create utility methods to further configure the mocked `JwtAuthenticationToken` built by `@WithMockAuthentication(value = JwtAuthenticationToken.class)`. Something like:
``` java
	@Test
	@WithMockAuthentication(authType = JwtAuthenticationToken.class, authorities = {"ROLE_AUTHORIZED_PERSONNEL"})
	public void test() throws Exception {
		polishMockedAuthentication(new HashMap<>());

		client.perform(get("/greet")).andExpect(content().string("Hello ch4mpy! You are granted with [ROLE_AUTHORIZED_PERSONNEL]."));
	}

	void polishMockedAuthentication(Map<String, Object> claims) {
		final JwtAuthenticationToken auth = (JwtAuthenticationToken) TestSecurityContextHolder.getContext().getAuthentication();
		final Jwt jwt = spy(Jwt.class);
		when(jwt.getClaims()).thenReturn(claims);
		// extra configuration for principal & credentials mocks
		when(auth.getPrincipal()).thenReturn(jwt);
		when(auth.getCredentials()).thenReturn(jwt);
		when(auth.getToken()).thenReturn(jwt);
	}
```

But I'm pretty sure both your production and test code would benefit using something a little more acctionable than `JwtAuthenticationToken`. Please give a try to `OidcIdToken` and let me know how it goes.

P.S.
Good catch to asign me the issue, I saw it rather quickly!

**Comment by wimdeblauwe** [comment]
I tried your first suggestion to create an `@WithMockJwtAuthenticationToken` and this seems to work fine:

```
import com.c4_soft.springaddons.security.oauth2.test.annotations.*;
import org.springframework.core.annotation.AliasFor;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.context.SecurityContext;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.oauth2.core.oidc.IdTokenClaimNames;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.security.oauth2.jwt.JwtClaimNames;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;
import org.springframework.security.test.context.support.TestExecutionEvent;
import org.springframework.security.test.context.support.WithSecurityContext;
import org.springframework.security.test.context.support.WithSecurityContextFactory;
import org.springframework.util.StringUtils;

import java.lang.annotation.*;
import java.net.MalformedURLException;
import java.time.Instant;
import java.util.Arrays;
import java.util.stream.Collectors;

@Target({ElementType.METHOD, ElementType.TYPE})
@Retention(RetentionPolicy.RUNTIME)
@Inherited
@Documented
@WithSecurityContext(factory = WithMockJwtAuthenticationToken.JwtAuthenticationTokenFactory.class)
public @interface WithMockJwtAuthenticationToken {
    @AliasFor("authorities")
    String[] value() default {};

    @AliasFor("value")
    String[] authorities() default {};

    IdTokenClaims id() default @IdTokenClaims();

    ClaimSet privateClaims() default @ClaimSet();

    @AliasFor(annotation = WithSecurityContext.class)
    TestExecutionEvent setupBefore() default TestExecutionEvent.TEST_METHOD;

    public final class JwtAuthenticationTokenFactory
            implements
            WithSecurityContextFactory<WithMockJwtAuthenticationToken> {

        @Override
        public SecurityContext createSecurityContext(WithMockJwtAuthenticationToken annotation) {
            final SecurityContext context = SecurityContextHolder.createEmptyContext();
            try {
                context.setAuthentication(authentication(annotation));
            } catch (MalformedURLException e) {
                throw new RuntimeException(e);
            }

            return context;
        }

        public JwtAuthenticationToken authentication(WithMockJwtAuthenticationToken annotation) throws MalformedURLException {
            Jwt.Builder builder = Jwt.withTokenValue("token")
                                     .header("alg", "none")
                                     .claim(JwtClaimNames.SUB, "user");

            addIdTokenClaimsToJwt(builder, annotation.id());
            addPrivateClaimsToJwt(builder, annotation.privateClaims());

            return new JwtAuthenticationToken(builder.build(),
                                              Arrays.stream(annotation.authorities()).map(SimpleGrantedAuthority::new).collect(Collectors.toList()));
        }

        private void addIdTokenClaimsToJwt(Jwt.Builder builder,
                                           IdTokenClaims tokenAnnotation) {
            if (StringUtils.hasLength(tokenAnnotation.authTime())) {
                builder.claim(IdTokenClaimNames.AUTH_TIME, Instant.parse(tokenAnnotation.authTime()));
            }
            if (StringUtils.hasLength(tokenAnnotation.exp())) {
                builder.expiresAt(Instant.parse(tokenAnnotation.exp()));
            }
            if (StringUtils.hasLength(tokenAnnotation.iat())) {
                builder.issuedAt(Instant.parse(tokenAnnotation.iat()));
            }
            if (StringUtils.hasText(tokenAnnotation.iss())) {
                builder.issuer(tokenAnnotation.iss());
            }
            if (StringUtils.hasLength(tokenAnnotation.jti())) {
                builder.claim(JwtClaimNames.JTI, tokenAnnotation.jti());
            }
            if (StringUtils.hasLength(tokenAnnotation.nbf())) {
                builder.notBefore(Instant.parse(tokenAnnotation.nbf()));
            }
            if (StringUtils.hasLength(tokenAnnotation.sessionState())) {
                builder.claim("session_state", tokenAnnotation.sessionState());
            }
            builder.subject(tokenAnnotation.sub())
                   .audience(Arrays.asList(tokenAnnotation.aud()))
                   .claim(IdTokenClaimNames.NONCE, tokenAnnotation.nonce())
                   .claim(IdTokenClaimNames.ACR, tokenAnnotation.acr())
                   .claim(IdTokenClaimNames.AMR, Arrays.asList(tokenAnnotation.amr()))
                   .claim(IdTokenClaimNames.AZP, tokenAnnotation.azp());
        }

        private void addPrivateClaimsToJwt(Jwt.Builder builder,
                                           ClaimSet privateClaims) {
            for (IntClaim claim : privateClaims.intClaims()) {
                builder.claim(claim.name(), claim.value());
            }
            for (var claim : privateClaims.longClaims()) {
                builder.claim(claim.name(), claim.value());
            }
            for (var claim : privateClaims.stringClaims()) {
                builder.claim(claim.name(), claim.value());
            }
            for (var claim : privateClaims.stringArrayClaims()) {
                builder.claim(claim.name(), claim.value());
            }
            for (var claim : privateClaims.jsonObjectClaims()) {
                builder.claim(claim.name(), JsonObjectClaim.Support.parse(claim));
            }
            for (var claim : privateClaims.jsonArrayClaims()) {
                builder.claim(claim.name(), JsonArrayClaim.Support.parse(claim));
            }
        }
    }
}
```

I can use it like this:

```
    @WithMockJwtAuthenticationToken(authorities = Roles.ADMIN,
            privateClaims = @ClaimSet(stringClaims = @StringClaim(name = "myPrivateClaim", value = "xxxxxx xxxxxx")))
```

If you think this would be useful for the library, you can copy that code, no problem.

I only kept `IdTokenClaims` and not `OidcStandardClaims` because I don't think that makes sense, but I don't know Jwt and Oidc deeply enough really to know if that is ok or not.

I am hesitant to introduce `OidcIdAuthenticationToken` in my production code, I would rather only depend on Spring Security things and just use your library in testing (for which it is amazing and I deeply thank you!).

PS: I did not assign the issue, don't know how that happened to be honest.

**Comment by ch4mpy** [comment]
@wimdeblauwe actually even `IdTokenClaims` don't make sense on plain JWT: an ID token is a JWT but JWT is not an ID token. 

I see 3 main levels of specification:
1. [JWT](https://datatracker.ietf.org/doc/html/rfc7519) (Json Web Token) which I assimilate to a serialised claim-set (that is a `Map<String, Object>`)
2. [OAuth](https://datatracker.ietf.org/doc/html/rfc6749) which builds on top of JWT, specifying a set of claims (you can expect some specific entries in the claim-set), some roles (user, client, resource-server and authorisation-server), and other stuff like authentication flows
3. [OIDC](https://openid.net/specs/openid-connect-core-1_0.html) (OpenID Connect) further specifying OAuth with more standard claims, standard resources on the authorisation-server (`.well-known/openid-configuration` from which clients can fetch configuration details), etc.

The set of [standard claims at step 1](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1) is very limited and of little use in apllication code (mostly relevant when decoding / validating a JWT).

Both [ID token](https://openid.net/specs/openid-connect-core-1_0.html#IDToken) and OIDC [standard](https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims) claims are specified in the document linked at step 3. As an OpenID token is a JWT, you can find JWT standard claims as a subset of ID token ones but the reverse is not true.

Now a days, any serious server delivering JWT is actually an OAuth 2 authorisation-server and is also complient with OpenID Connect.
Keycloak, Auth0, IdentityServer, etc, all are OIDC.
I think it is a big waste to stick with 7 claims, when the standard specifies so many more claims.

Look at this Jwt:
- why keeping the opaque token string after is parsed (constructor 1st argument)?
- why requiering `iat` and `eat` claims as constructor arg when also requiering the complete claims-set as 4th argument?
- why limiting to this 7 (usless) claims, when you know the token was issued by an OpenID authorisation server providing so much more (useful) claims?

I suggest that, before taking a definitive decision, you take time to compare source code between:
- `OidcId` and `Jwt` (see how lean the first is?) 
- `OidcIdAuthenticationToken` and `JwtAuthenticationToken` (noticed accessors return types? forsee how much more useful will the result of `.getPrincipal()` be?)

**Comment by ch4mpy** [comment]
@wimdeblauwe , I spent some time evaluating the opportunity to make your use-case easier (creating additional annotations for `Authentication` implementations I don't support).

This ended to quite important code shuffle, but I think it was worth it: I believe this lib is now eaiser to use (and maintain).

And as a bonus, I've implemented [`@WithMockJwtAuth`](https://github.com/ch4mpy/spring-addons/blob/master/spring-security-oauth2-test-addons/src/main/java/com/c4_soft/springaddons/security/oauth2/test/annotations/WithMockJwtAuth.java) . Plz have a look and give me your thoughts.

**Comment by ch4mpy** [comment]
1. `headers` is destinated to populate `Jwt` headers (constructor 4th argument). Your question reveal a problem in usabilty: it being typed `PrivateClaims` is pretty confusing. I renamed `PrivateClaims` to simply `Claims` which more concise and probably more expressive too.

2. My bad for exception you encountered. I hadn't tested with empty headers and there is an assert on non empty headers in `Jwt` constructor. Fixed : headers is now defaulted to the same value you used in your implementation: `Claims headers() default @Claims(stringClaims = @StringClaim(name = "alg", value = "none"));`
